src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestLocator.kt

```diff
@@ -55,21 +57,40 @@ object CargoTestLocator : SMTestLocator {
                     else -> null
                 }
                 if (sourceElement?.qualifiedName == qualifiedName) {
-                    add(PsiLocation.fromPsiElement(sourceElement))
+                    add(getLocation(project, sourceElement, lineNum))
                 }
             }
         }
     }
 
+    private fun getLocation(project: Project, element: PsiElement, lineNum: Int?) =
+        if (lineNum != null) {
+            object : PsiLocation<PsiElement>(project, element) {
+                override fun getOpenFileDescriptor(): OpenFileDescriptor? {
+                    val virtualFile = virtualFile
+                    val document = virtualFile?.document ?: return null
+                    val offset = if (lineNum < document.lineCount) {
+                        document.getLineStartOffset(lineNum)
+                    } else {
+                        psiElement.textOffset
+                    }
+                    return OpenFileDescriptor(project, virtualFile, offset)
+                }
+            }
+        } else {
+            PsiLocation.fromPsiElement(element)
+        }
+
     fun getTestUrl(name: String): String = "$TEST_PROTOCOL://$name"
 
     fun getTestUrl(function: RsQualifiedNamedElement): String =
         getTestUrl(function.qualifiedName ?: "")
 
-    private fun toQualifiedName(path: String): String {
+    private fun toQualifiedName(path: String): Pair<String, Int?> {
         val targetName = path.substringBefore(NAME_SEPARATOR).substringBeforeLast("-")
-        if (!path.contains(NAME_SEPARATOR)) return targetName
-        val qualifiedName = path.substringAfter(NAME_SEPARATOR)
-        return "$targetName$NAME_SEPARATOR$qualifiedName"
+        if (NAME_SEPARATOR !in path) return targetName to null
+        val qualifiedName = path.substringAfter(NAME_SEPARATOR).substringBefore("#")
+        val lineNum = if ("#" in path) path.substringAfterLast("#").toInt() - 1 else null
```

Comment:
Let's use `toIntOrNull` to avoid potential exception in future - as I understand cargo doesn't guarantee that output of doctest will remain the same

Reply:
Or, you can provide more strict rules while creation of location url. For example, create regexp (something like `\(line (\d+)\)`) that will extract line number from test output

Reply:
I'd personally prefer the second suggestion